### ISSUE TITLE
chore(deps): Migrate dependency management to `uv lock` + `uv sync` (hybrid PyTorch routing)

### ISSUE THREAD

**Issue by ktinubu** [tinaudio/synth-setter#1243]
# Migrate dependency management to `uv lock` + `uv sync` (hybrid PyTorch routing)

**Labels:** `infra`, `build`, `reproducibility`
**Type:** Tracking issue (multi-PR)

## Summary

We install dependencies imperatively today:

```bash
uv pip install --torch-backend ${TORCH_BACKEND} -r pyproject.toml --extra torch --extra dev
```

This produces **no lockfile**. Every install re-resolves the full dependency graph from our loose constraints, so transitive versions drift silently — between developer machines, between CI runs, and over time. For a project whose purpose is studying a specific, measurable effect, the failure mode is concrete: *we cannot reliably reconstruct the environment a published result was produced in.* A reviewer (or future us) asking "reproduce the numbers in Table 3" should not be blocked on "we don't know exactly which versions were installed that month."

This issue tracks three coupled changes:

1. Adopt a committed `uv.lock` as the single source of truth for installs.
2. Restructure `pyproject.toml` so PyTorch resolves correctly across all our hardware targets from that one lockfile.
3. Add a CI gate that verifies the lock stays valid on every platform we claim to support.

It also folds in two cheap dependency-hygiene fixes that naturally ride along with the relock.

---

## Background: why this is not a one-line change

The blocker is that **`--torch-backend` only works in the `uv pip` interface.** It does *not* work with `uv lock`, `uv sync`, or `uv run`. So we cannot keep our current auto-detect flow *and* gain a lockfile — adopting a lock requires moving the PyTorch backend routing into `pyproject.toml`.

### Is losing `--torch-backend=auto` a problem?

No, and arguably it's an improvement. `--torch-backend=auto` runs *only at install time*: it shells out to `nvidia-smi`, reads the CUDA driver version, and picks a matching wheel index. It has zero runtime presence — the installed `torch` behaves identically regardless of how the wheel was selected. No application code changes or breaks.

What we lose is the convenience of auto-detecting unknown/heterogeneous GPU drivers. What we *gain* is consistency: every GPU box gets the **same** pinned CUDA wheel from the lock, rather than a per-machine choice that varies with whatever driver each box happens to have. For reproducible research, a deliberate locked CUDA version is the correct direction, not a regression.

The one real obligation this transfers to us: the pinned CUDA wheel (e.g. `cu128`) must be compatible with the driver on every GPU box. CUDA wheels are forward-compatible with newer drivers, so any box whose driver supports ≥ the pinned toolkit version is fine — and drivers are almost always newer than the toolkit. The only failure case is a GPU box with a driver *too old* for the pinned version, where auto-detect would previously have silently picked an older wheel. See "Open questions" re: whether our fleet spans CUDA generations.

### The three targets and why markers alone can't do it

We must satisfy three backend targets from one universal lockfile:

| Target | Backend | How it's obtained |
|---|---|---|
| macOS (Apple Silicon) | MPS (Metal) | **There is no MPS index.** The macOS PyPI wheel already bundles Metal support. macOS must fall through to PyPI. |
| Linux/Windows + GPU | CUDA | PyTorch CUDA index (`download.pytorch.org/whl/cu128`) |
| GPU-less Linux CI that imports torch | CPU | PyTorch CPU index (`download.pytorch.org/whl/cpu`) |

There are two *independent* axes here:

- **macOS ↔ Linux** is a **platform fact**. Environment markers (`sys_platform`) express it perfectly.
- **CPU ↔ CUDA on the *same* OS** is a **choice**, not a platform fact. A marker cannot tell whether a given Linux box has a GPU. Only an explicit, selectable extra can express "this Linux install wants CPU torch."

Because we need *both* axes, neither markers-only nor extras-only is sufficient. The chosen design is the **hybrid**: platform markers for the OS axis + an isolated `cpu` extra for the same-OS CPU/CUDA axis. Alternatives considered and rejected:

- **Markers only** — clean, but cannot produce CPU torch on Linux. Breaks our GPU-less CI requirement.
- **Extras only (uv's documented default)** — handles CPU/CUDA, but has no MPS story: `--extra cpu` would route macOS to the PyTorch CPU index, which has no Metal, so Macs lose MPS.
- **Keep `--torch-backend`** — incompatible with locking; defeats the purpose.

---

## Proposed `pyproject.toml`

Changes: torch stack moves into regular `dependencies` (routed per-platform below); `cpu` becomes the only backend extra; `dev`/`docs` move into PEP 735 `[dependency-groups]`; index/source/conflict config added; `requires-python` gains an upper bound; `scipy` pin relaxed.

```toml
[project]
requires-python = ">=3.10,<3.14"   # upper bound stops the resolver chasing future Pythons with no torch wheels
dependencies = [
  # ... all existing non-torch deps stay verbatim ...

  # torch stack is now regular deps; backend is routed in [tool.uv.sources], not chosen here
  "torch>=2.0.0",
  "torchvision>=0.15.0",
  "torchaudio>=2.0.0",
  "lightning>=2.6.0",       # pure-python — must NOT appear in [tool.uv.sources]
  "torchmetrics>=0.11.4",   # pure-python — must NOT appear in [tool.uv.sources]
]

[project.optional-dependencies]
# The ONLY backend selector. Default (no extra) = CUDA on Linux/Win, PyPI/MPS on macOS.
# Pass `--extra cpu` ONLY on a GPU-less Linux CI runner that must import torch.
# On macOS the marker below excludes this extra entirely (it's a harmless no-op there).
cpu = ["torch>=2.0.0", "torchvision>=0.15.0", "torchaudio>=2.0.0"]

[dependency-groups]
# Development concerns — NOT published package features. `dev` is synced by default.
docs = ["mkdocs", "mkdocs-material", "mkdocstrings[python]", "griffe-pydantic"]
dev = [
  "hypothesis", "mutmut==3.5.*", "pre-commit", "pyright==1.1.408",
  "pytest", "pytest-benchmark", "pytest-cov", "pytest-xdist", "ruff",
  { include-group = "docs" },   # a plain `uv sync` installs dev AND docs
]

[tool.uv]
# `cpu` extra and the default CUDA path are mutually exclusive, so uv resolves each
# independently into the single lockfile rather than trying to satisfy both at once.
conflicts = [[{ extra = "cpu" }]]

[[tool.uv.index]]
name = "pytorch-cpu"
url = "https://download.pytorch.org/whl/cpu"
explicit = true                                   # don't search this index for unrelated packages

[[tool.uv.index]]
name = "pytorch-cu128"                             # CONFIRM this matches our GPU drivers
url = "https://download.pytorch.org/whl/cu128"
explicit = true

[tool.uv.sources]
# Resolution per platform:
#   macOS               -> no marker matches -> PyPI (MPS-capable wheel)
#   Linux/Win, no extra -> pytorch-cu128 (GPU default)
#   Linux/Win, --extra cpu -> pytorch-cpu (GPU-less CI)
# All three torch packages MUST be listed — if one is omitted it silently resolves
# from PyPI and pins a wheel that mismatches the others.
torch = [
  { index = "pytorch-cpu",   extra = "cpu", marker = "sys_platform == 'linux' or sys_platform == 'win32'" },
  { index = "pytorch-cu128",                marker = "sys_platform == 'linux' or sys_platform == 'win32'" },
]
torchvision = [
  { index = "pytorch-cpu",   extra = "cpu", marker = "sys_platform == 'linux' or sys_platform == 'win32'" },
  { index = "pytorch-cu128",                marker = "sys_platform == 'linux' or sys_platform == 'win32'" },
]
torchaudio = [
  { index = "pytorch-cpu",   extra = "cpu", marker = "sys_platform == 'linux' or sys_platform == 'win32'" },
  { index = "pytorch-cu128",                marker = "sys_platform == 'linux' or sys_platform == 'win32'" },
]
```

**Removed:** the `torch`, `dev`, and `all` extras. `all = ["synth-setter[torch,dev]"]` no longer makes sense — torch is a regular dep and dev/docs are default-installed groups. Replace any `synth-setter[all]` reference with `uv sync --all-groups --extra <backend>`. (If a *published* `all` extra is needed for non-uv consumers, raise it here and we'll add an extras-only aggregate.)

**Why dev/docs become groups, not extras:** they're private development concerns. As extras they leak our internal tooling choices into the package's published metadata; as PEP 735 groups they stay private *and* `dev` installs by default on a bare `uv sync`, which is what gets us close to a single everyday command.

---

## New command surface

```bash
uv sync                                              # mac (MPS) or GPU box (CUDA) — default, safe, dev+docs included
uv sync --extra cpu --no-default-groups --group dev  # GPU-less Linux CI that imports torch
uv sync --only-group dev                             # lint/type-check CI that does NOT import torch (no torch installed)
uv lock                                              # regenerate after dependency edits — run on a mac or GPU box only
```

Note GPU boxes and Macs use a **bare** `uv sync` (no `--extra`), which makes them immune to the resync flip described below.

---

## CI hygiene gate (this is what makes the lock trustworthy)

A universal lock is a *claim* about every platform it covers. It is generated on one machine and consumed on others, so CI must verify the claim holds. Two gates:

1. **`uv lock --check`** — fails if the committed lock is out of date with `pyproject.toml` (i.e. someone edited a dependency and forgot to relock). Run on a **matrix of macOS + Linux** so each platform's branch of the universal resolution is actually exercised, not just the author's.
2. **`uv sync --frozen`** in the real install steps — installs *exactly* the lock and errors on staleness, instead of silently re-resolving and masking drift.

Split the install matrix by our three targets so each is independently verified:

- macOS → `uv sync --frozen` (must get the MPS wheel)
- Linux GPU → `uv sync --frozen` (must get CUDA)
- Linux CPU CI → `uv sync --frozen --extra cpu --no-default-groups --group dev`

(A concrete GitHub Actions matrix can be added as a sub-task — see tasks.)

---

## Operational footgun: the resync flip

Two uv behaviors combine into a trap:

- Backend selection is **not sticky** — `--extra cpu` applies only to the command it's passed to; the next command has no memory of it.
- `uv sync` does **exact** syncing by default — it removes/replaces anything not matching the current command's resolution.

On Linux, CUDA is the default. So **any** Linux command without `--extra cpu` re-resolves torch to CUDA and exact-syncs the CPU wheel *out*, pulling multi-GB CUDA wheels onto a GPU-less box — frequently as a silent side effect of an unrelated action like `uv add pandas`. On a CPU-only box the flipped-in CUDA torch often still imports and runs (just finds no GPU), so the wrong, bloated environment can pass tests silently.

**Containment (do all three where practical):**
- Keep the CPU environment **ephemeral** — a fresh CI runner that runs one scripted `uv sync --extra cpu ...` and is destroyed has no "later command" to flip it.
- Set `UV_EXTRA=cpu` in the CPU-CI job so every uv invocation inherits the flag (removes the "remember the flag" risk entirely).
- **Never** run `uv add` or `uv lock` inside the `--extra cpu` environment. Do all dependency edits and relocks on a Mac or GPU box where the default resolution is the one we want, then commit the lock.

The flip is **Linux-only** in this design — it's the only platform where the `cpu` extra and the CUDA default coexist. Macs are immune.

---

## macOS-specific caveats (mostly inherent to MPS, not the config)

- **MPS is incomplete.** Not every op has an MPS kernel. Our stack has likely candidates: optimal transport (`POT`), scattering transforms (`kymatio`), custom losses/Sinkhorn. Hitting an unimplemented op raises a runtime "operator … not currently implemented for the MPS device" error (or silently falls back if `PYTORCH_ENABLE_MPS_FALLBACK=1` is set). Gate MPS-specific tests with the existing `mps` pytest marker.
- **Numerical divergence.** MPS results need not match CPU/CUDA bit-for-bit. Treat **CUDA as the source of truth for reported numbers**; don't be alarmed when a Mac-MPS run and a Linux-CUDA run differ. Worth a README note so a Mac collaborator doesn't file a spurious "results don't reproduce."
- **Intel Macs** silently get a CPU-only experience from the same wheel (no MPS). Expected, not a bug.
- **`--extra cpu` on a Mac is a misleading no-op** — the marker excludes it, so torch still comes from PyPI as the MPS wheel. Document: *Mac users run a bare `uv sync`, never a backend flag.*
- **arm64 source-build risk.** Watch `pedalboard`, `librosa`'s native deps, `h5py`/`hdf5plugin`, and the VST/Surge bits for source-build failures on Apple Silicon (a missing wheel triggers a from-source build during `uv sync` and looks like a uv error but isn't).

---

## Lock hygiene practices (ongoing)

- **Generate the lock somewhere that exercises every platform branch.** A lock made only on a Mac can encode a Linux resolution that was never actually validated (fine for static-metadata wheels; risky for varies-by-platform deps without static metadata). The macOS+Linux CI matrix running `uv lock --check` is what validates the whole claim.
- **Relock on a cadence** (`uv lock --upgrade` or targeted upgrades) and **review the lockfile diff like code** — especially torch/CUDA lines and native wheels. A lock diff can move a result.
- **If resolution fights us** (broad support surface, or one weak platform dragging everything backward), constrain `tool.uv.environments` to our concrete targets rather than the full platform Cartesian product.

---

## Tasks

- [ ] **Confirm CUDA target** (`cu128` vs `cu126`/`cu130`) against our GPU boxes' driver versions
- [ ] Restructure `pyproject.toml` per above (torch → deps; `cpu` extra; dev/docs → groups; indexes/sources/conflicts)
- [ ] Remove `torch`/`dev`/`all` extras; update any `synth-setter[all]` references → `uv sync --all-groups --extra <backend>`
- [ ] Add `requires-python` upper bound (`<3.14`)
- [ ] Relax `scipy==1.14.1` → `>=1.14,<1.15` (hard patch-pin risks backtracking against torch/librosa/kymatio); verify nothing depended on the exact pin
- [ ] Generate `uv.lock` on a Mac or GPU box and commit it
- [ ] Add CI matrix (macOS + Linux) running `uv lock --check`
- [ ] Switch CI install steps to `uv sync --frozen`, split by the three targets
- [ ] Update the skypilot bare-runner install in `.github/workflows/test-skypilot-debug.yml` (the `pyproject.toml` comment flags this must stay in sync)
- [ ] Set `UV_EXTRA=cpu` in the CPU-CI job, or confirm that environment is ephemeral
- [ ] README/CONTRIBUTING: document the command surface + "Mac users: bare `uv sync`, never a backend flag" + "CUDA is source of truth for reported numbers"
- [ ] Gate any MPS-incompatible tests behind the `mps` marker
- [ ] Establish a relock cadence (`uv lock --upgrade`) with lockfile diffs reviewed in PR

## Open questions

- **Does our GPU fleet span multiple CUDA generations?** If some boxes need an older toolkit than others, a single pinned `cu128` can't serve all of them. Fix is a second CUDA extra (e.g. `cu118`) routed the same way as `cpu`, letting each box pick — not reverting to `--torch-backend`.
- **Is the CPU-CI runner fresh per-run or cached/persistent?** Determines whether the resync flip is fully designed out by ephemerality or needs the `UV_EXTRA` guard.

## Out of scope

- Adding extra CUDA generations (only if the fleet turns out to span them — see open questions)
- Migrating interactive/dev workflows to `uv run`
- Any change to the non-torch dependency set beyond the scipy pin
